builtin/providers/aws/resource_aws_dms_endpoint.go

```diff
@@ -0,0 +1,334 @@
+package aws
+
+import (
+	"log"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/private/waiter"
+	dms "github.com/aws/aws-sdk-go/service/databasemigrationservice"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsDmsEndpoint() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsDmsEndpointCreate,
+		Read:   resourceAwsDmsEndpointRead,
+		Update: resourceAwsDmsEndpointUpdate,
+		Delete: resourceAwsDmsEndpointDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"certificate_arn": {
+				Type:         schema.TypeString,
+				Computed:     true,
+				Optional:     true,
+				ValidateFunc: validateArn,
+			},
+			"database_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
+			},
+			"endpoint_arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"endpoint_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"endpoint_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"source",
+					"target",
+				}, false),
+			},
+			"engine_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				// Valid Values: mysql, oracle, postgres, mariadb, aurora, redshift, sybase, sqlserver
+			},
+			"extra_connection_attributes": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
+			},
+			"kms_key_arn": {
+				Type:         schema.TypeString,
+				Computed:     true,
+				Optional:     true,
+				ValidateFunc: validateArn,
+			},
+			"password": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Sensitive: true,
+			},
+			"port": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"server_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"ssl_mode": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"none",
+					"require",
+					"verify-ca",
+					"verify-full",
+				}, false),
+			},
+			"tags": {
+				Type:     schema.TypeMap,
+				Optional: true,
+			},
+			"username": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceAwsDmsEndpointCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dmsconn
+
+	request := &dms.CreateEndpointInput{
+		EndpointIdentifier: aws.String(d.Get("endpoint_id").(string)),
+		EndpointType:       aws.String(d.Get("endpoint_type").(string)),
+		EngineName:         aws.String(d.Get("engine_name").(string)),
+		Password:           aws.String(d.Get("password").(string)),
```

Comment:
The API allows database_name, username, and password to be empty strings. I changed these to be required since it doesn't make much sense to set these to an empty string.